### PR TITLE
Move _raise_if_errors after recv in AsyncVectorEnv

### PR DESCRIPTION
**Pull request by tristandeleu**
There is a race condition in `AsyncVectorEnv`, most likely due to `_raise_if_errors` being called before the output of the workers is received: if the worker raises an error after `_raise_if_errors` is called, but before `pipe.recv()`, then the overall result contains spurious `None`, which cannot be iterated over anymore (see https://github.com/openai/gym/issues/1546#issuecomment-504585415).

### CODE REVIEW

File: gym/vector/async_vector_env.py [hunk 1]
```diff
@@ -276,7 +275,7 @@ def close(self, timeout=None, terminate=False):
                     pipe.send(('close', None))
             for pipe in self.parent_pipes:
                 if not pipe.closed:
-                    pipe.recv()
+                    _, success = pipe.recv()
```
Comment (christopherhesse) [inline gym/vector/async_vector_env.py]: Was this value supposed to be used anywhere?

Comment (christopherhesse) [reply gym/vector/async_vector_env.py]: Seems like it should be used or the change reverted.

**Comment by christopherhesse** [comment]
Thanks for fixing this!

File: gym/vector/async_vector_env.py (hunk 1 above)
Comment (tristandeleu) [reply gym/vector/async_vector_env.py]: Fixed it, this was not used anywhere

**Comment by christopherhesse** [comment]
This says it fixes #1542, but I think #1542 mentions two issues, and this only solves the second one, is that correct @tristandeleu?

**Comment by tristandeleu** [comment]
This actually should fix both. The first one was an issue with `pipe` being `None` in `_poll`, which is also part of the fix in this PR. The second one is handled by the new "success" flag returned by the workers.

**Comment by christopherhesse** [comment]
I ran the script there and it still has this error:

```
ERROR: Received the following error from Worker-1: AssertionError: Cannot call env.step() before calling reset()
ERROR: Shutting down Worker-1.
ERROR: Received the following error from Worker-2: AssertionError: Cannot call env.step() before calling reset()
ERROR: Shutting down Worker-2.
ERROR: Received the following error from Worker-0: AssertionError: Cannot call env.step() before calling reset()
ERROR: Shutting down Worker-0.
ERROR: Raising the last exception back to the main process.
Traceback (most recent call last):
  File "test.py", line 4, in <module>
    obs, rews, dones, infos = venv.step(venv.action_space.sample())
  File "/Users/csh/agi/gym/gym/vector/vector_env.py", line 54, in step
    return self.step_wait()
  File "/Users/csh/agi/gym/gym/vector/async_vector_env.py", line 228, in step_wait
    self._raise_if_errors(successes)
  File "/Users/csh/agi/gym/gym/vector/async_vector_env.py", line 333, in _raise_if_errors
    raise exctype(value)
AssertionError: Cannot call env.step() before calling reset()
Exception ignored in: <function AsyncVectorEnv.__del__ at 0x10d741ae8>
Traceback (most recent call last):
  File "/Users/csh/agi/gym/gym/vector/async_vector_env.py", line 338, in __del__
  File "/Users/csh/agi/gym/gym/vector/async_vector_env.py", line 262, in close
  File "/Users/csh/agi/gym/gym/logger.py", line 30, in warn
  File "/Users/csh/agi/gym/gym/utils/colorize.py", line 27, in colorize
ModuleNotFoundError: import of six halted; None in sys.modules
```

This isn't the sole fault of vectorenv I think, but it looks like it's using logger during close, and logger does a late import which fails because the interpreter is shutting down.

**Comment by christopherhesse** [comment]
Oh sorry about that, I said 1542 but I actually meant this issue: https://github.com/openai/gym/issues/1546

**Comment by tristandeleu** [comment]
Yes for #1546 it only fixes the second issue in the comment. For the first issue (and this one) I also think it is unrelated to `AsyncVectorEnv`. One fix would be to remove the logger from `close`, but this seems like a weak patch.

**Comment by christopherhesse** [comment]
I also checked out this branch against https://github.com/openai/gym/issues/1542 and still get an odd error there:

```
Traceback (most recent call last):
  File "test.py", line 16, in <module>
    main()
  File "test.py", line 12, in main
    obs, rews, dones, infos = venv.step(np.array([venv.action_space.sample() for _ in range(venv.num_envs)]))
  File "/Users/csh/agi/gym/gym/vector/vector_env.py", line 54, in step
    return self.step_wait()
  File "/Users/csh/agi/gym/gym/vector/async_vector_env.py", line 228, in step_wait
    self._raise_if_errors(successes)
  File "/Users/csh/agi/gym/gym/vector/async_vector_env.py", line 333, in _raise_if_errors
    raise exctype(value)
TypeError: only size-1 arrays can be converted to Python scalars
/Users/csh/agi/gym/gym/logger.py:30: UserWarning: WARN: Calling `close` while waiting for a pending call to `step` to complete.
  warnings.warn(colorize('%s: %s'%('WARN', msg % args), 'yellow'))
Exception ignored in: <function AsyncVectorEnv.__del__ at 0x117229ae8>
Traceback (most recent call last):
  File "/Users/csh/agi/gym/gym/vector/async_vector_env.py", line 338, in __del__
    self.close(terminate=True)
  File "/Users/csh/agi/gym/gym/vector/async_vector_env.py", line 281, in close
    pipe.close()
AttributeError: 'NoneType' object has no attribute 'close'
```

**Comment by christopherhesse** [comment]
I'll split out 1546 then

**Comment by christopherhesse** [comment]
Great, thanks!  I'll wait for tests to pass then merge this.

**Pull request merged**